syz-manager/covfilter.go

```diff
@@ -0,0 +1,215 @@
+			return "", err
+		}
+	}
+
+	covFilter := CoverFilter{
+		weightedPCs:    make(map[uint32]float32),
+		target:         target,
+		bitmapFilename: cfg.Workdir + "/" + "syz-cover-bitmap",
+	}
+	err = covFilter.initWeightedPCs(filesRegexp, funcsRegexp, rawPCs)
+	if err != nil {
+		return "", err
+	}
+	err = covFilter.createBitmap()
+	if err != nil {
+		return "", err
+	}
+	return covFilter.bitmapFilename, nil
+}
+
+func (covFilter *CoverFilter) initWeightedPCs(filesRegexp, funcsRegexp []regexp.Regexp, rawPCsFiles []string) error {
+	for _, f := range rawPCsFiles {
+		rawFile, err := os.Open(f)
+		if err != nil {
+			return fmt.Errorf("failed to open raw PCs file: %v", err)
+		}
+		for {
+			var encode uint64
+			_, err := fmt.Fscanf(rawFile, "0x%x\n", &encode)
+			if err != nil {
+				break
+			}
+			pc := uint32(encode & 0xffffffff)
+			weight := float32((encode >> 32) & 0xffff)
+			covFilter.weightedPCs[pc] = weight
+		}
+		rawFile.Close()
+	}
+
+	enabledFuncs := make(map[string]bool)
+	enabledFiles := make(map[string]bool)
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		if reportGenerator == nil {
+			return fmt.Errorf("reportGenerator used without initialization")
+		}
+		pcs := make([]uint64, 0)
+		symNames, cuNames, symPCs := reportGenerator.GetSymbolsInfo()
+		for i, symName := range symNames {
+			for _, r := range funcsRegexp {
+				if ok := r.MatchString(symName); ok {
+					enabledFuncs[symName] = true
+					pcs = append(pcs, symPCs[i]...)
+				}
+			}
+			for _, r := range filesRegexp {
+				if ok := r.MatchString(cuNames[i]); ok {
+					enabledFuncs[symName] = true
+					enabledFiles[cuNames[i]] = true
+					pcs = append(pcs, symPCs[i]...)
+				}
+			}
+		}
+		for _, pc := range pcs {
+			covFilter.weightedPCs[uint32(pc)] = 1.0
+		}
+	}
+	for funcs := range enabledFuncs {
+		log.Logf(1, "enabled kernel function: %s", funcs)
+	}
+	for files := range enabledFiles {
+		log.Logf(1, "enabled kernel file: %s", files)
+	}
+	return nil
+}
+
+func getRegexps(regexpStrings []string) ([]regexp.Regexp, error) {
+	var regexps []regexp.Regexp
+	for _, rs := range regexpStrings {
+		r, err := regexp.Compile(rs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compile regexp: %v", err)
+		}
+		regexps = append(regexps, *r)
+	}
+	return regexps, nil
+}
+
+func (covFilter *CoverFilter) createBitmap() error {
+	covFilter.detectRegion()
+	if covFilter.pcSize > 0 {
+		log.Logf(0, "coverage filter from 0x%x to 0x%x, size 0x%x",
+			covFilter.pcStart, covFilter.pcEnd, covFilter.pcSize)
+	} else {
+		return fmt.Errorf("coverage filter is enabled but nothing will be filtered")
+	}
+
+	bitmapFile, err := os.OpenFile(covFilter.bitmapFilename, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open or create bitmap: %s", err)
+	}
+	defer bitmapFile.Close()
+	bitmap := covFilter.bitmapBytes()
+	_, err = bitmapFile.Write(bitmap)
```

Comment:
It seems that this function can be reduced to:
```
return osutil.WriteFile(covFilter.bitmapFilename, covFilter.bitmapBytes())
```
at which point it does not worth to have it as a separate function.